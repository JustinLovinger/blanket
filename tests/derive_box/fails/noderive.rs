pub trait Counter {
    fn increment(&mut self);
}

#[derive(Default)]
struct AtomicCounter {
    count: u8,
}

impl Counter for AtomicCounter {
    fn increment(&mut self) {
        self.count += 1;
    }
}

struct CounterWrapper<C: Counter> {
    inner: C,
}

impl<C: Counter> From<C> for CounterWrapper<C> {
    fn from(inner: C) -> Self {
        Self { inner }
    }
}

fn main() {
    // counter wrapper should be able to wrap AtomicCounter
    let counter = AtomicCounter::default();
    let mut wrapper_by_value = CounterWrapper::from(counter);
    assert_eq!(wrapper_by_value.inner.count, 0);
    wrapper_by_value.inner.increment();
    assert_eq!(wrapper_by_value.inner.count, 1);
    // but this will fail because no implementation was derived
    let counter = AtomicCounter::default();
    let mut wrapper_by_box = CounterWrapper::from(Box::new(counter));
    assert_eq!(wrapper_by_box.inner.count, 0);
    wrapper_by_box.inner.increment();
    assert_eq!(wrapper_by_box.inner.count, 1);
}
