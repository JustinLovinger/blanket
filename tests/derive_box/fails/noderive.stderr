error[E0277]: the trait bound `std::boxed::Box<AtomicCounter>: Counter` is not satisfied
  --> $DIR/noderive.rs:35:51
   |
16 | struct CounterWrapper<C: Counter> {
   |                          ------- required by this bound in `CounterWrapper`
...
35 |     let mut wrapper_by_box = CounterWrapper::from(Box::new(counter));
   |                                                   ^^^^^^^^^^^^^^^^^ the trait `Counter` is not implemented for `std::boxed::Box<AtomicCounter>`

error[E0277]: the trait bound `std::boxed::Box<AtomicCounter>: Counter` is not satisfied
  --> $DIR/noderive.rs:35:30
   |
16 | struct CounterWrapper<C: Counter> {
   |                          ------- required by this bound in `CounterWrapper`
...
35 |     let mut wrapper_by_box = CounterWrapper::from(Box::new(counter));
   |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Counter` is not implemented for `std::boxed::Box<AtomicCounter>`
